### ISSUE TITLE
Bug: `TSError: Unexpected token. Did you mean `{'>'}` or `&gt;`?`

### ISSUE THREAD

**Issue by Akronae** [typescript-eslint/typescript-eslint#10377]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

parser

### Repro Code

```TypeScript
import { parseForESLint } from "@typescript-eslint/parser";

let code = `
import {type Context, useContext} from "react";

export const useSafeContext = <T>(context: Context<T>) => {
    const value = useContext(context);
    if (!value) {
        throw new Error(context.displayName as string);
    }
    return value;
};
`
const parsed = parseForESLint(code, {
    filePath: 'test.tsx',
    jsx: true
})
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/<rule-name>": ["error", ...<options>],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

I expect the parse function not to throw errors

### Actual Result

```
/home/alex/prog/a/node_modules/@typescript-eslint/typescript-estree/dist/node-utils.js:619
    return new TSError(message, ast.fileName, { start, end });
           ^

TSError: Unexpected token. Did you mean `{'>'}` or `&gt;`?
    at createError (/home/alex/prog/a/node_modules/@typescript-eslint/typescript-estree/dist/node-utils.js:619:12)
    at convertError (/home/alex/prog/a/node_modules/@typescript-eslint/typescript-estree/dist/convert.js:41:41)
    at astConverter (/home/alex/prog/a/node_modules/@typescript-eslint/typescript-estree/dist/ast-converter.js:15:42)
    at parseAndGenerateServices (/home/alex/prog/a/node_modules/@typescript-eslint/typescript-estree/dist/parser.js:162:66)
    at parseForESLint (/home/alex/prog/a/node_modules/@typescript-eslint/parser/dist/parser.js:97:80)
    at file:///home/alex/prog/a/node_modules/eslint-plugin-exception-handling/dist/src/utils/parse.js:14:16
    at ModuleJob.run (node:internal/modules/esm/module_job:262:25)
    at async onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:485:26)
    at async asyncRunEntryPointWithESMLoader (node:internal/modules/run_main:109:5) {
  fileName: '/home/alex/prog/a/test.tsx',
  location: {
    start: { line: 4, column: 56, offset: 106 },
    end: { line: 4, column: 56, offset: 106 }
  }
}

Node.js v22.5.1
```

### Additional Info

The issue is caused by a generic argument `<T>` before the lambda

Working fine:
```js
import { parseForESLint } from "@typescript-eslint/parser";

let code = `
import {type Context, useContext} from "react";

export const useSafeContext = (context: Context<T>) => {
    const value = useContext(context);
    if (!value) {
        throw new Error(context.displayName as string);
    }
    return value;
};
`
const parsed = parseForESLint(code, {
    filePath: 'test.tsx',
    jsx: true
})

```

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/parser`            | `7.9.0` |
| `node`                                 | `22.5.1` |
